// @wordpress/block-editor/build/components/block-styles/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _lodash = require("lodash");

var _classnames = _interopRequireDefault(require("classnames"));

var _compose = require("@wordpress/compose");

var _components = require("@wordpress/components");

var _previewPanel = _interopRequireDefault(require("./preview-panel"));

var _useStylesForBlock = _interopRequireDefault(require("./use-styles-for-block"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function BlockStylesPreviewPanelSlot(_ref) {
  let {
    scope
  } = _ref;
  return (0, _element.createElement)(_components.Slot, {
    name: `BlockStylesPreviewPanel/${scope}`
  });
}

function BlockStylesPreviewPanelFill(_ref2) {
  let {
    children,
    scope,
    ...props
  } = _ref2;
  return (0, _element.createElement)(_components.Fill, {
    name: `BlockStylesPreviewPanel/${scope}`
  }, (0, _element.createElement)("div", props, children));
} // Top position (in px) of the Block Styles container
// relative to the editor pane.
// The value is the equivalent of the container's right position.


const DEFAULT_POSITION_TOP = 16; // Block Styles component for the Settings Sidebar.

function BlockStyles(_ref3) {
  let {
    clientId,
    onSwitch = _lodash.noop,
    onHoverClassName = _lodash.noop,
    scope
  } = _ref3;
  const {
    onSelect,
    stylesToRender,
    activeStyle,
    genericPreviewBlock,
    className: previewClassName
  } = (0, _useStylesForBlock.default)({
    clientId,
    onSwitch
  });
  const [hoveredStyle, setHoveredStyle] = (0, _element.useState)(null);
  const [containerScrollTop, setContainerScrollTop] = (0, _element.useState)(0);
  const isMobileViewport = (0, _compose.useViewportMatch)('medium', '<');
  (0, _element.useLayoutEffect)(() => {
    const scrollContainer = document.querySelector('.interface-interface-skeleton__content');
    const scrollTop = (scrollContainer === null || scrollContainer === void 0 ? void 0 : scrollContainer.scrollTop) || 0;
    setContainerScrollTop(scrollTop + DEFAULT_POSITION_TOP);
  }, [hoveredStyle]);

  if (!stylesToRender || stylesToRender.length === 0) {
    return null;
  }

  const debouncedSetHoveredStyle = (0, _lodash.debounce)(setHoveredStyle, 250);

  const onSelectStylePreview = style => {
    onSelect(style);
    onHoverClassName(null);
    setHoveredStyle(null);
    debouncedSetHoveredStyle.cancel();
  };

  const styleItemHandler = item => {
    var _item$name;

    if (hoveredStyle === item) {
      debouncedSetHoveredStyle.cancel();
      return;
    }

    debouncedSetHoveredStyle(item);
    onHoverClassName((_item$name = item === null || item === void 0 ? void 0 : item.name) !== null && _item$name !== void 0 ? _item$name : null);
  };

  return (0, _element.createElement)("div", {
    className: "block-editor-block-styles"
  }, (0, _element.createElement)("div", {
    className: "block-editor-block-styles__variants"
  }, stylesToRender.map(style => {
    const buttonText = style.label || style.name;
    return (0, _element.createElement)(_components.Button, {
      className: (0, _classnames.default)('block-editor-block-styles__item', {
        'is-active': activeStyle.name === style.name
      }),
      key: style.name,
      variant: "secondary",
      label: buttonText,
      onMouseEnter: () => styleItemHandler(style),
      onFocus: () => styleItemHandler(style),
      onMouseLeave: () => styleItemHandler(null),
      onBlur: () => styleItemHandler(null),
      onClick: () => onSelectStylePreview(style),
      "aria-current": activeStyle.name === style.name
    }, (0, _element.createElement)(_components.__experimentalText, {
      as: "span",
      limit: 12,
      ellipsizeMode: "tail",
      className: "block-editor-block-styles__item-text",
      truncate: true
    }, buttonText));
  })), hoveredStyle && !isMobileViewport && (0, _element.createElement)(BlockStylesPreviewPanelFill, {
    scope: scope,
    className: "block-editor-block-styles__preview-panel",
    style: {
      top: containerScrollTop
    },
    onMouseLeave: () => styleItemHandler(null)
  }, (0, _element.createElement)(_previewPanel.default, {
    activeStyle: activeStyle,
    className: previewClassName,
    genericPreviewBlock: genericPreviewBlock,
    style: hoveredStyle
  })));
}

BlockStyles.Slot = BlockStylesPreviewPanelSlot;
var _default = BlockStyles;
exports.default = _default;
