// @wordpress/block-editor/build-module/components/colors-gradients/tools-panel-color-dropdown.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { Button, ColorIndicator, Dropdown, FlexItem, __experimentalHStack as HStack, __experimentalToolsPanelItem as ToolsPanelItem } from '@wordpress/components';
/**
 * Internal dependencies
 */

import ColorGradientControl from './control';
import useMultipleOriginColorsAndGradients from './use-multiple-origin-colors-and-gradients';
export default function ToolsPanelColorDropdown(_ref) {
  var _settings$gradientVal;

  let {
    settings,
    enableAlpha,
    ...otherProps
  } = _ref;
  const colorGradientSettings = useMultipleOriginColorsAndGradients();
  const controlSettings = { ...colorGradientSettings,
    clearable: false,
    enableAlpha,
    label: settings.label,
    onColorChange: settings.onColorChange,
    onGradientChange: settings.onGradientChange,
    colorValue: settings.colorValue,
    gradientValue: settings.gradientValue
  };
  const selectedColor = (_settings$gradientVal = settings.gradientValue) !== null && _settings$gradientVal !== void 0 ? _settings$gradientVal : settings.colorValue;
  return createElement(ToolsPanelItem, _extends({
    hasValue: settings.hasValue,
    label: settings.label,
    onDeselect: settings.onDeselect,
    isShownByDefault: settings.isShownByDefault,
    resetAllFilter: settings.resetAllFilter
  }, otherProps, {
    className: "block-editor-tools-panel-color-gradient-settings__item"
  }), createElement(Dropdown, {
    className: "block-editor-tools-panel-color-dropdown",
    contentClassName: "block-editor-panel-color-gradient-settings__dropdown-content",
    renderToggle: _ref2 => {
      let {
        isOpen,
        onToggle
      } = _ref2;
      return createElement(Button, {
        onClick: onToggle,
        "aria-expanded": isOpen,
        className: classnames('block-editor-panel-color-gradient-settings__dropdown', {
          'is-open': isOpen
        })
      }, createElement(HStack, {
        justify: "flex-start"
      }, createElement(ColorIndicator, {
        className: "block-editor-panel-color-gradient-settings__color-indicator",
        colorValue: selectedColor
      }), createElement(FlexItem, null, settings.label)));
    },
    renderContent: () => createElement(ColorGradientControl, _extends({
      showTitle: false,
      __experimentalHasMultipleOrigins: true,
      __experimentalIsRenderedInSidebar: true,
      enableAlpha: true
    }, controlSettings))
  }));
}
