// @wordpress/block-editor/build/components/block-popover/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlockPopover;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classnames = _interopRequireDefault(require("classnames"));

var _components = require("@wordpress/components");

var _dom = require("@wordpress/dom");

var _useBlockRefs = require("../block-list/use-block-props/use-block-refs");

var _usePopoverScroll = _interopRequireDefault(require("./use-popover-scroll"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function BlockPopover(_ref) {
  let {
    clientId,
    bottomClientId,
    children,
    __unstablePopoverSlot,
    __unstableContentRef,
    ...props
  } = _ref;
  const selectedElement = (0, _useBlockRefs.__unstableUseBlockElement)(clientId);
  const lastSelectedElement = (0, _useBlockRefs.__unstableUseBlockElement)(bottomClientId !== null && bottomClientId !== void 0 ? bottomClientId : clientId);
  const popoverScrollRef = (0, _usePopoverScroll.default)(__unstableContentRef);

  if (!selectedElement || bottomClientId && !lastSelectedElement) {
    return null;
  }

  const anchorRef = {
    top: selectedElement,
    bottom: lastSelectedElement
  };
  const {
    ownerDocument
  } = selectedElement;
  const stickyBoundaryElement = ownerDocument.defaultView.frameElement || (0, _dom.getScrollContainer)(selectedElement) || ownerDocument.body;
  return (0, _element.createElement)(_components.Popover, (0, _extends2.default)({
    ref: popoverScrollRef,
    noArrow: true,
    animate: false,
    position: "top right left",
    focusOnMount: false,
    anchorRef: anchorRef,
    __unstableStickyBoundaryElement: stickyBoundaryElement // Render in the old slot if needed for backward compatibility,
    // otherwise render in place (not in the the default popover slot).
    ,
    __unstableSlotName: __unstablePopoverSlot || null,
    __unstableBoundaryParent: true // Observe movement for block animations (especially horizontal).
    ,
    __unstableObserveElement: selectedElement,
    shouldAnchorIncludePadding: true // Used to safeguard sticky position behavior against cases where it would permanently
    // obscure specific sections of a block.
    ,
    __unstableEditorCanvasWrapper: __unstableContentRef === null || __unstableContentRef === void 0 ? void 0 : __unstableContentRef.current
  }, props, {
    className: (0, _classnames.default)('block-editor-block-popover', props.className)
  }), children);
}
